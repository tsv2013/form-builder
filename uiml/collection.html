<div class='collection' data-bind="">
    <div class="collection__header"
        data-bind="click: function () { collapsed(!collapsed()); }, css: { 'collection__header--collapsed': collapsed }">
        <span data-bind="text: title"></span>
    </div>
    <div class="collection__content" data-bind="css: { 'collection__content--hidden': collapsed }">
        <span class="bf-item-menu__button" data-bind="click: add">+ Add</span>
        <!-- ko foreach: items -->
        <object-editor params="context: $data, type: $parent.elementType"></object-editor>
        <div class="bf-item-menu__button" data-bind="click: function() { $parent.remove($data); }">- Remove</div>
        <!-- /ko -->
    </div>
</div>