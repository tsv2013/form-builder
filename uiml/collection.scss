collection-editor {
  display: block;
  position: relative;
  min-width: 200px;
}

.collection__header {
  cursor: pointer;
}

.collection__content--hidden {
  display: none;
}

.collection__item {
  padding: 7px 0;
  // background-color: rgb(175, 175, 175);
  border-left: 2px solid rgb(175, 175, 175);
  border-right: 2px solid rgb(175, 175, 175);
}
.collection__item:not(:first-child) {
  margin-top: 7px;
}

.collection__item-menu {
  padding-top: 7px;
}