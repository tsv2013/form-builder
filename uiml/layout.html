<!-- ko foreach: elements -->
<layout-item params="element: $data" data-bind="style: {  }"></layout-item>
<!-- /ko -->
<!-- ko if: elements().length === 0 -->
<placeholder-item params="element: formElement"></placeholder-item>
<!-- /ko -->
