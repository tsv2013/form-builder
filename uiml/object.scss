object-editor {
  display: block;
  position: relative;
  min-width: 300px;
}
