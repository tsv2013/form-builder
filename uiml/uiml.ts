export interface IPart {
    id?: string;
    partclass: string;
    parts?: IPart[];
    cssClasses?: string;
    [others: string]: any;
}

export interface IStyle {
    partname?: string;
    partclass?: string;
    name: string;
    value: string;
}

export interface IUMLInterface {
    parts: IPart[];
    styles?: IStyle[];
}

export interface IProperty {
    id: string;
    mapsto: string;
}

export interface IClass {
    id: string;
    mapsto: string;
    dproperties?: IProperty[];
}

export interface IPeer {
    peername: string;
    id: string;
    dclasses: IClass[];
    wloaders: { [widget: string]: () => Promise<any> };
}

export interface IUML {
    uimlinterface: IUMLInterface;
    peers: IPeer[];
}

export let htmlUiRendererPeers = [
    <IPeer>{
        peername: "presentation",
        id: "HTMLUIRenderer",
        dclasses: [
            { id: "container", mapsto: "ContainerRenderer" },
            { id: "input", mapsto: "InputRenderer" },
            { id: "label", mapsto: "LabelRenderer" }
        ],
        wloaders: <{ [widget: string]: () => Promise<any> }>{}
    }
];
