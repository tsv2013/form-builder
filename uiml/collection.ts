import * as ko from "knockout";
import { htmlUiRendererPeers, render } from "./index";

import "./collection.scss";

var collectionEditorTemplate = require("text-loader!./collection.html");

export class CollectionWidgetViewModel {
    private _itemsSubscription: ko.Subscription;
    private readonly model: any;

    constructor(private options: { data: string, elementType: string }, public context: any) {
        this.model = context.$data;
        let data = this.model[options.data] || [];
        this.items(data);
        this._itemsSubscription = this.items.subscribe(items => this.model[this.options.data] = items);
    }

    collapsed = ko.observable(true);

    items = ko.observableArray();

    get title() {
        return `${this.collapsed()?'>':'v'} ${this.options.data} (${this.items().length})`;
    }

    get elementType() {
        return this.options.elementType;
    }

    add = (collectionWidget: CollectionWidgetViewModel) => {
        // TODO: pass creator function as a parameter
        this.items.push(this.model.create(this.elementType));
    }

    remove = (item: any) => {
        this.items.remove(item);
    }

    dispose() {
        this._itemsSubscription.dispose();
    }
}

ko.components.register("collection-editor", {
    viewModel: {
        createViewModel: function (params, componentInfo) {
            return new CollectionWidgetViewModel(params.options, params.context);
        }
    },
    template: collectionEditorTemplate
});

// Widget registration in UIML rendering peers
//noinspection TypeScriptUnresolvedFunction
htmlUiRendererPeers[0].dclasses.push({
    id: "collection-editor",
    mapsto: "ComponentRenderer"
});