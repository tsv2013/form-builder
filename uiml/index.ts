import { IUML, IUMLInterface, IPart, htmlUiRendererPeers } from "./uiml";
import { UIMLWrapper } from "./uiml-wrapper";

export * from "./uiml";
export * from "./renderer";
export * from "./panel";
export * from "./layout";
export * from "./object";
export * from "./collection";

declare var config: IUML;

const uimlWrapper = new UIMLWrapper(typeof config !== "undefined" && config || { uimlinterface: <IUMLInterface>{}, peers: htmlUiRendererPeers });

export function render(part: IPart, container?: HTMLElement) {
    const renderer = uimlWrapper.createRenderer(part);
    return renderer.render(container);
}

export function refresh() {
    uimlWrapper.refresh();
}