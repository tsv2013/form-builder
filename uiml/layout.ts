import * as ko from "knockout";

import { htmlUiRendererPeers } from "./uiml";

import "./layout.scss";

var layoutTemplate = require("text-loader!./layout.html");

export class LayoutViewModel {

}

ko.components.register("layout", {
    viewModel: {
        createViewModel: function (params, componentInfo) {
            // ko.utils.domNodeDisposal.addDisposeCallback(componentInfo.element, () => {
            //     subscription.dispose();
            // });
            var panelFormElement = params.context.$parents[2];
            var layoutFormElement = panelFormElement.elements()[0];
            return {
                formElement: layoutFormElement,
                elements: ko.computed(() => {
                    return layoutFormElement.elements();
                })
            };
        }
    },
    template: layoutTemplate
});

// Widget registration in UIML rendering peers
//noinspection TypeScriptUnresolvedFunction
htmlUiRendererPeers[0].dclasses.push({
    id: "layout",
    mapsto: "ComponentRenderer"
});
