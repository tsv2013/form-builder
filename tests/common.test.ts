import { render } from "../uiml";

test("render label", () => {
    var result = render({ partclass: "label" });
    expect(result).toMatchSnapshot();
    expect(result.tagName).toBe("LABEL");
});

test("render input", () => {
    var result = render({ partclass: "input", data: "name" });
    expect(result).toMatchSnapshot();
    expect(result.tagName).toBe("INPUT");
});
