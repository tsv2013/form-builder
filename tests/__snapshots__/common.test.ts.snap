// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render input 1`] = `
<input
  class="form-control "
  data-bind="value: $data.name, valueUpdate: 'keyup'"
  placeholder=""
/>
`;

exports[`render label 1`] = `
<label
  class=" "
/>
`;
